sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -1703,4 +1774,498 @@ public void onCompletion(RecordMetadata metadata, Exception exception) {
     throw new RuntimeException(String.format(
         "Could not extract the Kafka Deserializer type from %s", deserializer));
   }
+
+  //////////////////////////////////  Exactly-Once Sink   \\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\
+
+  /**
+   * Exactly-once sink transform.
+   */
+  private static class EOSWrite<K, V> extends PTransform<PCollection<KV<K, V>>, PCollection<Void>> {
+    //
+    // Dataflow ensures at-least once processing for side effects like sinks. In order to provide
+    // exactly-once semantics, a sink needs to be idempotent or it should avoid writing records
+    // that have already been written. This snk does the latter. All the the records are ordered
+    // across a fixed number of shards and records in each shard are written in order. It drops
+    // any records that are already written and buffers those arriving out of order.
```

Comment:
<!--new_thread; commit:27f622c4e2c0fabf68d43da4e4aa381f71f82365; resolved:0-->
We should not make this specific to Dataflow, though it might make assumptions that not all runners support right now about GroupByKey (I know that at least Dataflow and Spark do support these assumptions today)